improver/calibration/rainforest_calibration.py

```diff
@@ -126,6 +144,433 @@ def __init__(self, model_config_dict: dict, threads: int):
+            - forecast_cube with realization coordinate added if absent
+
+        Raises:
+            ValueError:
+                if feature/forecast variables have inconsistent dimension coordinates
+                (excluding realization dimension), or if feature/forecast variables have
+                different length realization coordinate over cubes containing a realization
+                coordinate.
+        """
+        combined_cubes = CubeList(list([*feature_cubes, forecast_cube]))
+
+        # Compare feature cube coordinates, raise error if dim-coords don't match
+        compare_feature_coords = compare_coords(
+            combined_cubes, ignored_coords=["realization"]
+        )
+        misaligned_dim_coords = [
+            coord_info["coord"]
+            for misaligned_coords in compare_feature_coords
+            for coord, coord_info in misaligned_coords.items()
+            if coord_info["data_dims"] is not None
+        ]
+        if misaligned_dim_coords:
+            raise ValueError(
+                f"Dimension coords do not match between: {misaligned_dim_coords}"
+            )
+
+        # Compare realization coordinates across cubes where present;
+        # raise error if realization coordinates don't match, otherwise set
+        # common_realization_coord to broadcast over.
+        realization_coords = {
+            variable.name(): variable.coords("realization")
+            for variable in combined_cubes
+            if variable.coords("realization")
+        }
+        if not realization_coords:
+            # Case I: realization_coords is empty. Add single realization dim to all cubes.
+            common_realization_coord = DimCoord(
+                [0], standard_name="realization", units=1, var_name="realization"
+            )
+        else:
+            # Case II: realization_coords is not empty.
+            variables_with_realization = list(realization_coords.keys())
+            sample_variable = variables_with_realization[0]
+            sample_realization = realization_coords[sample_variable][0]
+            misaligned_realizations = [
+                feature
+                for feature in variables_with_realization[1:]
+                if realization_coords[feature][0] != sample_realization
+            ]
+            if misaligned_realizations:
+                misaligned_realizations.append(sample_variable)
+                raise ValueError(
+                    f"Realization coords  do not match between: {misaligned_realizations}"
+                )
+            common_realization_coord = sample_realization
+
+        # Add realization coord to cubes where absent by broadcasting along this dimension
+        aligned_cubes = CubeList()
+        for i_cube, cube in enumerate(combined_cubes):
+            if not cube.coords("realization"):
+                cube = combined_cubes[i_cube]
+                expanded_cube = add_coordinate(
```

Comment:
Originally I had implemented a function to add a new dimension coordinate by broadcasting the data along the new coordinate, but dropped it as this functionality already exists through `add_coordinate` (see https://github.com/metoppv/improver/pull/1721/commits/5a5ca9d466881748d42475d2e5e0fcb2a0c57368).

However, the points you raise are valid. I've done as suggested and found the following:
* `add_coordinate` does ok on small to moderate sized cubes, but is slow for large cubes (took ~10s to broadcast `(10000, 10000)` to `(50, 10000, 10000)`)
* comparing to the function I had in previous commit, this implementation is much faster (took ~1 ms to broadcast `(10000, 10000)` to `(50, 10000, 10000)`) 

I also had a look around the codebase to see how similar operations on cubes are carried out. Firstly, `add_coordinate` is not used outside of testing, so given this and the performance check I'm thinking the current approach is not the right way to go. In regards to similar operations, most broadcasting (that I could find) tended to work directly on the ndarray and was typically bespoke to the situation where the broadcasting was required.

The only comparable usage was in ensemble_calibration which calls a function:https://github.com/metoppv/improver/blob/b18f795db5f79930820d323e0122de0630dda0e1/improver/calibration/utilities.py#L378
to broadcast cube data across `time` coordinate for those that lacking this coordinate. Rather than returning a CubeList, it returns a list of ndarrays; so the usage is similar, but not identical to our usage.

I think there is some value in having a function that broadcasts a cube across a coordinate, and given my implementation is more efficient I think I'll add this to the codebase (into `utilities.cube_manipulation`, or somewhere similar).

As this function will essentially duplicate the functionality of `add_coordinate`, is it worth swapping this out (this would be better placed in a separate PR as there are a lot of instances of this function being referenced), or leave it knowing that this function is used only for testing?

Reply:
>As this function will essentially duplicate the functionality of `add_coordinate`, is it worth swapping this out (this would be better placed in a separate PR as there are a lot of instances of this function being referenced), or leave it knowing that this function is used only for testing?

The current interface I've put forwards for `add_coordinate_to_cube` differs from that of `add_coordinates`, and the internals differ somewhat so I think it's probably best to leave as is.